charts/helmet/templates/_helpers.tpl

```diff
@@ -8,3 +8,36 @@ Create the name of the service account to use
 {{- default "default" .Values.serviceAccount.name }}
 {{- end }}
 {{- end }}
+
+
+{{/*
+Render an array of env variables. The input can be a map or a slice.
+Values can be templates using the "common.tplvalues.render" helper, but changes to scope are not processed.
+Usage:
+{{ include "helmet.toEnvArray" ( dict "envVars" .Values.envVars "context" $ ) }}
+*/}}
+{{- define "helmet.toEnvArray" -}}
+{{- if kindIs "map" .envVars }}
+{{- range $key, $val := .envVars }}
+- name: {{ $key }}
+{{- if kindIs "string" $val }}
+  value: {{ include "common.tplvalues.render" (dict "value" $val "context" $.context) }}
+{{- else if kindIs "map" $val }}
+{{- if $val.value }}
+  value: {{ include "common.tplvalues.render" (dict "value" $val.value "context" $.context) }}
+{{- else if $val.valueFrom }}
+  valueFrom: {{ include "common.tplvalues.render" (dict "value" $val.valueFrom "context" $.context) | nindent 4 }}
+{{- end }}
```

Comment:
In this section, you don't need to check `value` or `valueFrom`. If `$val` is a `map`, then simply return it, like this:

```suggestion
  {{ include "common.tplvalues.render" (dict "value" $val "context" $.context) }}
```

The output of the suggested code should be similar to your code but simplified.

Reply:
Good point, like with the slice, I didn't realize `"common.tplvalues.render"` worked recursively. I have some extra changes for this suggestion though: 
- instead of the leading indent we need to use `| indent 2`, otherwise the valuesFrom isn't indented properly:
  ```yaml
  - name: BASIC_VALUE
    value: basic value
  - name: CONFIGMAP
    valueFrom:
    configMapRef:          # <- indent 2 fixes these multiline entries
      key: my-key          # <-
      name: my-config-map  # <-
  ```
- I didn't want a name key to be rendered, since the name comes from the map's key in this case, so I added `(omit $val "name")` to the template to prevent this case from rendering like: 
  ```text
  # file: values.yaml                       | # Undesirable rendered output (deployment.yaml)
  envVars:                                  |   ...
    BASIC_NAME_IGNORED:                     |   env:
      name: should not see this             |    - name: BASIC_NAME_IGNORED
      value: basic value name ignored       |      name: should not see this
                                            |      value: basic value name ignored
  ```
  - I wanted this to be the output for the case where a name is specified: 
    ```yaml
    # Desired rendered output (deployment.yaml)
      ...
      env:
        - name: BASIC_NAME_IGNORED
          value: basic value name ignored
    ```